### PR TITLE
✨ [feature] Add support for `Shock` sensor

### PR DESCRIPTION
**Pull request by pbuckley4192**
Hi,

First off, thanks for writing this App. It saved my IQPanel 2 from the scrap heap!

I noticed my setup uses "shock" sensors which werent supported by the App.

Error Seen: 
`2024-01-05 14:15:43.717826 WARNING qolsys_panel: sensor of unknown type: {'id': '170-7272', 'type': 'Shock', 'name': 'Upstairs Shock Detector', 'group': 'shock', 'status': 'Closed', 'state': '0', 'zone_id': 14, 'zone_physical_type': 12, 'zone_alarm_type': 0, 'zone_type': 107, 'partition_id': 0}
`

I've added the code needed locally and tested it against my setup. It functions as expected.

More than open to feedback!

### CODE REVIEW

File: apps/qolsysgw/mqtt/updater.py [hunk 1]
```diff
@@ -444,6 +445,7 @@ class MqttWrapperQolsysSensor(MqttWrapper):
         QolsysSensorKeypad: 'safety',
         QolsysSensorMotion: 'motion',
         QolsysSensorSiren: 'safety',
+        QolsysSensorShock: 'safety',
```
Comment (XaF) [inline apps/qolsysgw/mqtt/updater.py:448]: Probably more `vibration` ?

File: apps/qolsysgw/qolsys/sensors.py [hunk 2]
```diff
@@ -374,3 +374,8 @@ class QolsysSensorDoorbell(QolsysSensor):
     @classmethod
     def from_json(cls, data, partition, common=None):
         return cls.from_json_subclass('Doorbell', data, partition, common)
+
```
Comment (XaF) [inline apps/qolsysgw/qolsys/sensors.py:377]: Double space between classes or linter will complain

```suggestion


```

**Review by XaF — COMMENTED** [review]
Hello!

Thank you for contributing!
Could you please add the relevant test and fixtures too?

You should have the following files modified by this PR to ensure we're testing the right stuff:
- [x] apps/qolsysgw/mqtt/updater.py
- [x] apps/qolsysgw/qolsys/sensors.py
- [ ] tests/end-to-end/test_qolsysgw.py
- [ ] tests/integration/test_qolsys_events.py
- [ ] tests/mock_modules/testutils/fixtures_data.py

**Comment by pbuckley4192** [comment]
I'll take care of these on Monday! Thanks for the feedback!

**Comment by pbuckley4192** [comment]
All files updated.

 * [x]  apps/qolsysgw/mqtt/updater.py
 * [x]  apps/qolsysgw/qolsys/sensors.py
 * [x]  tests/end-to-end/test_qolsysgw.py
 * [x]  tests/integration/test_qolsys_events.py
 * [x]  tests/mock_modules/testutils/fixtures_data.py

**Note:** I found a missing dependency in requirements-tools.txt. Took the chance to add it.

File: tests/requirements-tools.txt [hunk 3]
```diff
@@ -3,3 +3,4 @@ pytest
 pytest-cov
 pytest-clarity
 pytest-subtests
+copier
```
Comment (XaF) [inline tests/requirements-tools.txt:6]: This is already in https://github.com/XaF/qolsysgw/blob/main/tests/end-to-end/requirements.txt (for end-to-end testing), this is not required for other tests!

File: apps/qolsysgw/mqtt/updater.py [hunk 4]
```diff
@@ -444,6 +445,7 @@ class MqttWrapperQolsysSensor(MqttWrapper):
         QolsysSensorKeypad: 'safety',
         QolsysSensorMotion: 'motion',
         QolsysSensorSiren: 'safety',
+        QolsysSensorShock: 'vibration',
```
Comment (XaF) [inline apps/qolsysgw/mqtt/updater.py:448]:
```suggestion
        QolsysSensorShock: 'vibration',
        QolsysSensorSiren: 'safety',
```

for alphabetic order

File: tests/integration/test_qolsys_events.py [hunk 5]
```diff
@@ -805,6 +806,29 @@ async def test_integration_event_info_summary_initializes_all_entities(self):
                 expected_enabled_by_default=False,
             )
 
+        with self.subTest(msg='Sensor 20090 is properly configured'):
+            sensor20090 = partition1.zone(20090)
+            self.assertEqual(QolsysSensorShock, sensor20090.__class__)
+            self.assertEqual('002-0090', sensor20090.id)
+            self.assertEqual('My Shock Sensor', sensor20090.name)
+            self.assertEqual('shock', sensor20090.group)
```
Comment (XaF) [inline tests/integration/test_qolsys_events.py:814]:
```suggestion
            self.assertEqual('Shock', sensor20090.group)
```

**Review by XaF — COMMENTED** [review]
Another few comments, and there's two tests failing.
One of them related to the expected number of sensors in the tests, since this adds one :)

**Comment by pbuckley4192** [comment]
Made the required changes, thanks for the feedback

**Comment by XaF** [comment]
Still some failing tests, are you able to fix that? :)

**Comment by XaF** [comment]
It seems to be failing at the e2e tests now :(

It is because the group here is written with a capital `S`: https://github.com/XaF/qolsysgw/pull/143/files#diff-94f35fef3f4d094a51239e392ab0d740b53bfb8f3462e6ec246865437e0799c5R382
While the end-to-end tests are written with a lowercase `s`.

According to your message in the PR description, I believe it's supposed to be a lowercase `s` for `shock`, which means you will need to update it at the line I quoted, as well as in the unit tests! Thanks a lot for your time in contributing this :)

File: apps/qolsysgw/qolsys/sensors.py [hunk 6]
```diff
@@ -374,3 +374,9 @@ class QolsysSensorDoorbell(QolsysSensor):
...
+
+class QolsysSensorShock(QolsysSensor):
+    @classmethod
+    def from_json(cls, data, partition, common=None):
+        return cls.from_json_subclass('Shock', data, partition, common)
```
Comment (XaF) [inline apps/qolsysgw/qolsys/sensors.py:382]:
```suggestion
        return cls.from_json_subclass('shock', data, partition, common)
```

File: tests/integration/test_qolsys_events.py [hunk 7]
```diff
@@ -805,6 +806,29 @@ async def test_integration_event_info_summary_initializes_all_entities(self):
                 expected_enabled_by_default=False,
             )
 
+        with self.subTest(msg='Sensor 20090 is properly configured'):
+            sensor20090 = partition1.zone(20090)
+            self.assertEqual(QolsysSensorShock, sensor20090.__class__)
+            self.assertEqual('002-0090', sensor20090.id)
+            self.assertEqual('My Shock Sensor', sensor20090.name)
+            self.assertEqual('Shock', sensor20090.group)
```
Comment (XaF) [inline tests/integration/test_qolsys_events.py:814]:
```suggestion
            self.assertEqual('shock', sensor20090.group)
```

File: tests/mock_modules/testutils/fixtures_data.py [hunk 8]
```diff
@@ -323,6 +323,19 @@ def get_summary(secure_arm=False, partition_ids=None,
                         'zone_type': 20,
                         'partition_id': 1,
                     },
+                    {
+                        'id': '002-0090',
+                        'type': 'Shock',
+                        'name': 'My Shock Sensor',
+                        'group': 'Shock',
```
Comment (XaF) [inline tests/mock_modules/testutils/fixtures_data.py:330]:
```suggestion
                        'group': 'shock',
```

File: tests/integration/test_qolsys_events.py [hunk 9]
```diff
@@ -22,6 +22,7 @@
 from qolsys.sensors import QolsysSensorPanelGlassBreak
 from qolsys.sensors import QolsysSensorPanelMotion
 from qolsys.sensors import QolsysSensorSiren
+from qolsys.sensors import QolsysSensorShock
```
Comment (XaF) [inline tests/integration/test_qolsys_events.py:25]:
```suggestion
from qolsys.sensors import QolsysSensorShock
from qolsys.sensors import QolsysSensorSiren
```

File: apps/qolsysgw/qolsys/sensors.py (hunk 6 above)
Comment (XaF) [reply apps/qolsysgw/qolsys/sensors.py:382]: Eh, I was wrong for this one, the type was correct, just not the group!

**Review by XaF — APPROVED** [review]

**Comment by XaF** [comment]
Ok, we're good! 🎉 
Thanks for the PR!

**Pull request merged**